device-connectors/src/testflinger_device_connectors/devices/zapper/__init__.py

```diff
@@ -51,24 +52,24 @@ class ZapperConnector(ABC, DefaultDevice):
     @catch(RecoveryError, 46)
     def provision(self, args):
         """Method called when the command is invoked."""
-        with open(args.config) as configfile:
-            config = yaml.safe_load(configfile)
-        testflinger_device_connectors.configure_logging(config)
+        with open(args.config, encoding="utf-8") as configfile:
+            self.config = yaml.safe_load(configfile)
+        with open(args.job_data, encoding="utf-8") as job_json:
+            self.job_data = json.load(job_json)
 
-        (api_args, api_kwargs) = self._validate_configuration(
-            args.config, args.job_data
-        )
+        testflinger_device_connectors.configure_logging(self.config)
 
         logger.info("BEGIN provision")
         logger.info("Provisioning device")
 
-        self._run(args.config["controller_host"], *api_args, **api_kwargs)
+        (api_args, api_kwargs) = self._validate_configuration()
+        self._run(self.config["control_host"], *api_args, **api_kwargs)
 
         logger.info("END provision")
 
     @abstractmethod
     def _validate_configuration(
-        self, config, job_data
+        self,
     ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
```

Comment:
If you're using Python > 3.5, you can use it like this (see [docs](https://docs.python.org/3/library/typing.html) for more):
```suggestion
    ) -> tuple[tuple[Any, ...], dict[str, Any]]:
```

Reply:
Mh unfortunately the lower bound for this syntax (TIL is called [Generic Alias Type](https://docs.python.org/3/library/stdtypes.html#generic-alias-type)) is 3.9. What you can do since 3.5 is writing [your own alias](https://docs.python.org/3/library/typing.html#type-aliases) like `type MyDict = dict[T, T]` which is basically typing.Dict (https://docs.python.org/3/library/typing.html#aliases-to-built-in-types).

Reply:
Oh, I see. We can deprecate it once we drop Python 3.8 support😌